src/database.js

```diff
@@ -42,17 +43,57 @@ exports.update = function (bill) {
 };
 
 exports.save = async function (watchlist) {
+	const writeFile = () => fs.writeFile(
+		`${databaseDirectory}${databaseFile}`,
+		JSON.stringify(database, null, '	')
+	).then(() => dirty = false);
+
+	const mkdirsAndWriteFile = async () => {
+		const dirParts = databaseDirectory.split('/').filter((d, j) => d.length > 0 || j === 0); // `j === 0` in case the start is at root
+		const joinDirParts = (i) => dirParts.filter((p, j) => j < i).join('/');
+		const mkdir = (i) => fs.mkdir(joinDirParts(i));
+		var i = dirParts.length;
+		var lastError;
+
+		for ( ; i > 0; i--) {
+			try {
+				await mkdir(i);
+				break;
+			} catch (err) {
+				lastError = err;
+				if (err.code === 'ENOENT') {
+					continue; // verbose for verbosity's sake
+				} else if (err.code === 'EEXIST') {
+					console.warn(`database.js:save: the directory ${joinDirParts(i)} already exists, retrying the next directories down in the tree (unexpected)`); // big warn
+					break;
+				} else {
+					console.warn(`database.js:save: unhandled error [1]\n${err}`); // big warn, but continue
+				}
+			}
+		}
+
+		if (i > 0) {
+			for (i++; i < dirParts.length + 1; i++) {
+				try {
+					await mkdir(i);
+				} catch (err) {
+					console.warn(`database.js:save: unhandled error [2]\n${err}`); // big warn, but continue
+				}
+			}
+			return writeFile();
+		} else {
+			console.warn(`database.js:save: could not make any of the directories in the tree ${databaseDirectory}`); // big warn
+			return Promise.reject(lastError); // for sanity
```

Comment:
why is promise.reject necessary? won't an error thrown here also cause the async function's promise to reject?

Reply:
I believe since we caught and handled the error in the async try-catch fashion that no rejection would be caused by the anonymous function. Farther down at line 94, we would want to know that there was an error that occurred during our fallback `mkdirsAndWriteFile` method.
In any case, now that you've reminded me of the recursive option for `fs.mkdir`, a lot of this will be removed, including returning this rejected Promise.